File: ios/exelbid_plugin/Sources/exelbid_plugin/Banner/BannerPlatformView.swift
import ExelBidSDK
import Flutter
import UIKit

final class BannerPlatformView: NSObject, FlutterPlatformView, FlutterStreamHandler {

    private let containerView: UIView
    private var bannerAd: EBBannerAd?
    private var eventSink: FlutterEventSink?
    private let eventChannel: FlutterEventChannel
    private let methodChannel: FlutterMethodChannel

    init(frame: CGRect, viewId: Int64, arguments: [String: Any],
         messenger: FlutterBinaryMessenger) {
        let container = UIView(frame: frame)
        container.backgroundColor = .clear
        self.containerView = container

        self.methodChannel = FlutterMethodChannel(
            name: ChannelNames.Banner.method(forViewId: viewId),
            binaryMessenger: messenger
        )
        self.eventChannel = FlutterEventChannel(
            name: ChannelNames.Banner.events(forViewId: viewId),
            binaryMessenger: messenger
        )

        super.init()

        self.eventChannel.setStreamHandler(self)
        self.methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
        setupBanner(arguments: arguments)
    }

    deinit {
        bannerAd?.stop()
        bannerAd?.removeFromSuperview()
        methodChannel.setMethodCallHandler(nil)
        eventChannel.setStreamHandler(nil)
    }

    // MARK: - FlutterPlatformView

    func view() -> UIView { containerView }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // MARK: - Method channel

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "load":
            bannerAd?.load()
        case "stop":
            bannerAd?.stop()
        default:
            result(FlutterMethodNotImplemented)
            return
        }
        result(nil)
    }

    // MARK: - Banner setup

    private func setupBanner(arguments: [String: Any]) {
        let adUnitId = arguments["adUnitId"] as? String ?? ""
        let width = arguments["width"] as? Double ?? 320
        let height = arguments["height"] as? Double ?? 50
        let autoRefresh = arguments["autoRefresh"] as? Bool ?? true
        let autoLoad = arguments["autoLoad"] as? Bool ?? true
        let fullWebView = arguments["fullWebView"] as? Bool ?? false
        let optionsMap = arguments["options"] as? [String: Any]

        let banner = EBBannerAd(adUnitId: adUnitId, size: CGSize(width: width, height: height))
        banner.options = AdOptionsMapper.decode(optionsMap)
        banner.autoRefresh = autoRefresh
        banner.fullWebView = fullWebView
        wireCallbacks(banner: banner)

        containerView.addSubview(banner)
        banner.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            banner.topAnchor.constraint(equalTo: containerView.topAnchor),
            banner.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
            banner.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            banner.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
        ])

        self.bannerAd = banner
        if autoLoad {
            banner.load()
        }
    }

    private func wireCallbacks(banner: EBBannerAd) {
        banner.onLoad = { [weak self] in
            self?.send(["event": "onLoad"])
        }
        banner.onFailureBlock = { [weak self] error in
            self?.send([
                "event": "onFail",
                "error": AdErrorMapper.encode(error as NSError),
            ])
        }
        banner.onClick = { [weak self] in
            self?.send(["event": "onClick"])
        }
        banner.onLeaveApp = { [weak self] in
            self?.send(["event": "onLeaveApp"])
        }
        banner.onClickFinish = { [weak self] in
            self?.send(["event": "onClickFinish"])
        }
    }

    private func send(_ payload: [String: Any]) {
        // Event channel sinks must be invoked on the main thread.
        if Thread.isMainThread {
            eventSink?(payload)
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.eventSink?(payload)
            }
        }
    }
}

File: android/src/main/kotlin/com/exelbid/flutter/banner/BannerPlatformView.kt
package com.exelbid.flutter.banner

import android.content.Context
import android.view.View
import android.widget.FrameLayout
import com.exelbid.flutter.ChannelNames
import com.exelbid.flutter.mappers.AdErrorMapper
import com.exelbid.flutter.mappers.AdOptionsMapper
import com.onnuridmc.exelbid.ExelBidAdView
import com.onnuridmc.exelbid.common.ExelBidError
import com.onnuridmc.exelbid.common.OnBannerAdListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/**
 * PlatformView for a standalone (non-mediated) ExelBid banner — direct
 * [ExelBidAdView]. Honors `autoRefresh`/`autoLoad`/`fullWebView` and exposes
 * `load`/`stop` over the per-view MethodChannel.
 *
 * `OnBannerAdListener` only reports loaded/failed/clicked, so onLeaveApp /
 * onClickFinish are not emitted on Android.
 */
class BannerPlatformView(
    context: Context,
    viewId: Long,
    args: Map<String, Any?>?,
    messenger: BinaryMessenger,
) : PlatformView, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val container = FrameLayout(context)
    private val methodChannel = MethodChannel(messenger, ChannelNames.Banner.method(viewId))
    private val eventChannel = EventChannel(messenger, ChannelNames.Banner.events(viewId))

    private var sink: EventChannel.EventSink? = null
    private val pending = ArrayList<Map<String, Any?>>()

    private val banner: ExelBidAdView = ExelBidAdView(context)

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)

        val adUnitId = args?.get("adUnitId") as? String ?: ""
        val autoRefresh = args?.get("autoRefresh") as? Boolean ?: true
        val autoLoad = args?.get("autoLoad") as? Boolean ?: true
        val fullWebView = args?.get("fullWebView") as? Boolean ?: false

        @Suppress("UNCHECKED_CAST")
        val options = args?.get("options") as? Map<String, Any?>

        banner.setAdUnitId(adUnitId)
        banner.setFullWebView(fullWebView)
        AdOptionsMapper.apply(banner, options)
        if (!autoRefresh) banner.setAutoRefreshDisable()
        banner.setAdListener(object : OnBannerAdListener {
            override fun onAdLoaded() {
                emit(mapOf("event" to "onLoad"))
            }

            override fun onAdFailed(errorCode: ExelBidError?, statusCode: Int) {
                emit(
                    mapOf(
                        "event" to "onFail",
                        "error" to AdErrorMapper.encode(
                            0,
                            errorCode?.errorMessage ?: "Banner failed ($statusCode)",
                        ),
                    ),
                )
            }

            override fun onAdClicked() {
                emit(mapOf("event" to "onClick"))
            }
        })

        container.addView(
            banner,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT,
            ),
        )

        if (autoLoad) banner.loadAd()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "load" -> {
                banner.loadAd()
                result.success(null)
            }
            "stop" -> {
                banner.setAutoRefreshDisable()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun emit(event: Map<String, Any?>) {
        val s = sink
        if (s != null) s.success(event) else pending.add(event)
    }

    override fun getView(): View = container

    override fun dispose() {
        banner.destroy()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        sink = null
        pending.clear()
        container.removeAllViews()
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        if (events != null && pending.isNotEmpty()) {
            pending.forEach { events.success(it) }
            pending.clear()
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
